### DISCUSSION TITLE
Should autoload functions and completion be zcompiled?

### DISCUSSION THREAD

**Discussion by rieje** [jandamm/zgenom#154] [Ideas]
Disclaimer: new to zsh.

Should autoload functions and completion be zcompiled? I took a glance at zcompile implementations between some plugin managers and have a few questions:

* It seems zgenom zcompiles the most appropriate set of files (tests for zsh-parsable, ignores symlinks). It doesn't include autoload functions (unless it has the suffix `.zsh` but glancing over at random dotilfes, the naming scheme seems to be just the name of the function without the suffix--I could be wrong). It also doesn't include completion files. Should these be zcompiled or handled?

* Would it make sense to compile multiple files into one .zwc file to reduce the number of files every shell init needs to read (or one per "module"? zcompile is usually instant on personal config so I'm not sure if there are introduced downsides. Checking if any one of the files in a .zwc digest was updated to update the digest might not be as straightforward though. This idea is not new, I came across it as a feature request from a project I can't remember.

* Is there a general rule for when to use zcompile -M vs -R? I think I understand everything in the manpage except `-R`'s:

> When  compiling  scripts  instead  of  autoloadable  functions, it is often desirable to use this option; otherwise the whole file, including the code to define functions which have already  been  defined,  will remain mapped, consequently wasting memory.

* Zcompiling zsh builtin functions and scripts at /usr/share/zsh/{functions,scripts,site-functions} would be beneficial, right? I've never seen these covered.

Some of these may not be appropriate for a plugin manager to do but I'm still curious if there are any thoughts.

Much appreciated.

P.S. The only files zgenom missed were .zsh-theme files (I'm not sure if that's a standard naming scheme but I've seem them in other plugin managers, e.g. for files like powerlevel10k.zsh-theme).

**Comment by jandamm** [comment]
- At least `zgenom`s own autoloaded functions start with `#!/usr/bin/env zsh` and will be compiled.
- From my understanding of compiled zsh files it works like this: `zsh` needs to execute/source a file at path `a/b/c.zsh`. It will then check if there is a `a/b/c.zsh.zwc` which isn't older than `a/b/c/.zsh` and will use the `zwc`. Otherwise it will use `a/b/c.zsh`. So to create a "module" the source files would need to be merged.
- I don't know and let the system decide. `zgenom` only uses `-U` which makes it safer to use when the user has overwritten a command with an alias.
- `zgenom` will not compile anything out of it's scope. Meaning itself and managed plugins (and the zcompdump if it knows where it is). You can however use `zgenom compile` to compile your dotfiles or other files (at your own risk).

Compiling `.zsh-theme`s as well seems reasonable 👍
